spherical_splines/SHTOOLS/man/man1/makegeoidgrid.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAKEGEOIDGRID 1"
.TH MAKEGEOIDGRID 1 "2009-08-18" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "MakeGeoidGrid"
.IX Header "MakeGeoidGrid"
.IP "MakeGeoidGrid \-" 17
.IX Item "MakeGeoidGrid -"
Create a global map of the geoid.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 MakeGeoidGrid (" 27
.IX Item "SUBROUTINE MakeGeoidGrid ("
\&\s-1GEOID\s0, \s-1CILM\s0, \s-1LMAX\s0, R0POT, \s-1GM\s0, \s-1POTREF\s0, \s-1OMEGA\s0, R, \s-1GRIDTYPE\s0, \s-1ORDER\s0, \s-1NLAT\s0, \s-1NLONG\s0, \s-1INTERVAL\s0, \s-1LMAX_CALC\s0, A, F )
.RS 4
.IP "REAL*8" 18
.IX Item "REAL*8"
\&\s-1GEOID\s0(\s-1NLAT\s0, \s-1NLONG\s0), \s-1CILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0), R0POT, \s-1GM\s0, \s-1POTREF\s0, \s-1OMEGA\s0, R
.IP "\s-1INTEGER\s0" 18
.IX Item "INTEGER"
\&\s-1LMAX\s0, \s-1GRIDTYPE\s0, \s-1ORDER\s0, \s-1NLAT\s0, \s-1NLONG\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 18
.IX Item "REAL*8, OPTIONAL"
\&\s-1INTERVAL\s0, A, F
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 18
.IX Item "INTEGER, OPTIONAL"
\&\s-1LMAX_CALC\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIMakeGeoidGrid\fR will create a global map of the geoid, accurate to either first, second, or third order, using the method described in Wieczorek (2007; equation 19\-20). The algorithm expands the potential in a Taylor series on a spherical interface of radius R, and computes the height above this interface to the potential \s-1POTREF\s0 exactly from the linear, quadratic, or cubic equation at each grid point. If the optional parameters A and F are specified, the geoid height will be referenced to a flattened ellipsoid with semi-major axis A and flattening F. The pseudo-rotational potential is explicitly accounted for by specifying the angular rotation rate \s-1OMEGA\s0 of the planet.
.PP
It should be noted that this geoid calculation is only strictly exact when the radius R lies above the maximum radius of the planet. Furthermore, the geoid is only strictly valid when it lies above the surface of the planet (it is necessary to know the density structure of the planet when calculating the potential below the surface).
.PP
The geoid can be computed on one of four different grids: (1) a Gauss-Legendre quadrature grid (see \fIMakeGridGLQ\fR), (2) A N by N equally sampled grid (see \fIMakeGridDH\fR), (3) an N by 2N equally spaced grid (see \fIMakeGridDH\fR), or (4) A 2D Cartesian grid (see \fIMakeGrid2D\fR),.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1GEOID\s0" 10
.IX Item "GEOID"
(output) REAL*8, \s-1DIMENSION\s0(\s-1NLAT\s0, \s-1NLONG\s0)
.Sp
A global grid of the height to the potential \s-1POTREF\s0 above a sphere of radius R (or above a flattened ellipsoid if both A and F are specified). The number of latitude and longitude points depends upon \s-1GRIDTYPE:\s0 (1) \s-1LMAX+1\s0 by 2*LMAX + 1, (2) 2*LMAX+2 by 2*LMAX+2, (3) 2*LMAX+2 by 4*LMAX+4, or (4) 180/INTERVAL + 1 by 360/INTERVAL + 1.
.IP "\s-1CILM\s0" 10
.IX Item "CILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The real spherical harmonic coefficients (geodesy normalized) of the gravitational potential referenced to a spherical interface of radius R0POT.
.IP "\s-1LMAX\s0" 10
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic degree of the gravitational-potential coefficients. For GRIDTYPEs 1, 2 and 3, this determines the number of latitudinal and longitudinal samples.
.IP "R0POT" 10
.IX Item "R0POT"
(input) REAL*8
.Sp
The reference radius of the spherical harmonic coefficients.
.IP "\s-1GM\s0" 10
.IX Item "GM"
(input) REAL*8
.Sp
The product of the gravitational constant and mass of the planet.
.IP "\s-1OMEGA\s0" 10
.IX Item "OMEGA"
(input) REAL*8
.Sp
The angular rotation rate of the planet.
.IP "R" 10
.IX Item "R"
(input) REAL*8
.Sp
The radius of the reference sphere that the Taylor expansion of the potential is performed on. If A and F are not specified, the geoid height will be referenced to this spherical interface.
.IP "\s-1GRIDTYPE\s0" 10
.IX Item "GRIDTYPE"
(input) \s-1INTEGER\s0
.Sp
The output grid is (1) a Gauss-Legendre quadrature grid whose grid nodes are determined by \s-1LMAX\s0, (2) an equally sampled N by N grid used with the Driscoll and Healy (1994) sampling theorem, (3) ar a similar N by 2N grid that is oversampled in longitude, or (4) a 2D Cartesian grid with latitudinal and longitudinal spacing given by \s-1INTERVAL\s0.
.IP "\s-1ORDER\s0" 10
.IX Item "ORDER"
(input) \s-1INTEGER\s0
.Sp
The order of the Taylor series expansion of the potential about the reference radius R. This can be either 1, 2, or 3.
.IP "\s-1NLAT\s0" 10
.IX Item "NLAT"
(output) \s-1INTEGER\s0
.Sp
The number of latitudinal samples.
.IP "\s-1NLONG\s0" 10
.IX Item "NLONG"
(output) \s-1INTEGER\s0
.Sp
The number of longitudinal samples.
.IP "\s-1INTERVAL\s0" 10
.IX Item "INTERVAL"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The latitudinal and longitudinal spacing of the output grid for \s-1GRIDTYPE\s0 is 4.
.IP "\s-1LMAX_CALC\s0" 10
.IX Item "LMAX_CALC"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
The maximum degree used in evaluating the spherical harmonic coefficients.
.IP "A" 10
.IX Item "A"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The semi-major axis of the flattened ellipsoid that the output grid \s-1GEOID\s0 is referenced to. The optional parameter F must also be specified.
.IP "F" 10
.IX Item "F"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The flattening (R_equator \- R_pole)/R_equator of the reference ellipsoid. The optional parameter A (i.e., R_equator) must be specified.
.SH "NOTES"
.IX Header "NOTES"
This routine uses geodesy 4\-pi normalized spherical harmonics that exclude the Condon-Shortley phase. This can not be modified.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fImakegrid2d\fR\|(1), \fImakegridglq\fR\|(1), \fImakegriddh\fR\|(1)
.PP
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "REFERENCES"
.IX Header "REFERENCES"
Driscoll, J.R. and D.M. Healy, Computing Fourier transforms and convolutions on the 2\-sphere, \fIAdv. Appl. Math.\fR, 15, 202\-250, 1994.
.PP
Wieczorek, M. A. Gravity and topography of the terrestrial planets, \fITreatise on Geophysics\fR, 10, 165\-206, 2007.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005\-2009 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/MakeGeoidGrid.f95
subroutine MakeGeoidGrid(geoid, cilm, lmax, r0pot, GM, PotRef, omega, r, gridtype, &
			order, nlat, nlong, interval, lmax_calc, a, f)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	This subrouine will calculate the height to an equipotential surface with
!	respect to a spherical reference radius R for a body rotating with angular rotation
!	rate OMEGA. This method uses a first-, second-, or third-order Taylor series expansion 
!	of the potential, evaluated at radius R. The output gridype is specificied by GRIDTYPE, 
!	which can be either Cartesian 2D, Guass-Legendre quadrature, or Driscoll and Healy. 
!	If the optional parameters A and F are specified, the geoid will be referenced to a 
!	flattened ellispoid with semimajor axis A and flattening F.
!	
!	Note that this routine is only strictly valid when the geoid is above the surface!
!	To calculated the height of the geoid when it is below the surface, one would
!	need to know the density structure of the planet. Furthermore, the calculation of the 
!	potential (and its derivatives) is only strictly valid when R lies above the maximum radius
!	of the planet.
!
!	Latitude is geocentric latitude.
!
!	Calling Parameters
!		IN
!			cilm		Spherical harmonic potential coefficients.
!			lmax		Maximum spherical harmonic degree. For gridtype=1, this
!					is simply the maximum degree used in evaluating the 
!					spherical harmonic coefficients. For the other grids,
!					this sets the spacing of the output grid.
!			r0pot		Reference radius of the potential coefficients.
!			GM		GM associated with the potential coefficients.
!			PotRef		Reference potential of the geoid.
!			omega		Angular rotation rate used when calculating the geoid
!					on in the rotating coordinate system.
!			r		Reference sperical radius to calculate the geoid heights.
!			gridtype	1 = Gauss-Legendre quadrature grid corresponding to LMAX.
!					2 = N by N Driscoll and Healy grid corresponding to LMAX.
!					3 = N by 2N Driscoll and Healy grid corresponding to LMAX.
!					4 = 2D Cartesian using MakeGrid2D.
!			order 		Order of the Taylor expansion. Either 1, 2, or 3.
!		OUT
!			geoid		Gridded values of the geoid, in meters, referenced to the spherical
!					radius R.
!			nlat, nlong	Number of latitude and longitude points for the grid.
!		OPTIONAL, IN
!			interval	Grid spacing of the output grid in DEGREES. Used only when
!					GRIDTYPE = 1.
!			lmax_calc	For GRIDTYPE 2, 3, and 4, this specifies the maximum spherical
!					harmonic degree to evaluate the function to.
!			a		Semimajor axis of the reference flattened ellipsoid.
!			f		Flattening of the reference ellipsoid.
!			
!			
!
!	Dependencies :  MakeGrid2D, MakeGridGLQ, MakeGridDH, NGLQSH, PreCompute
!
!	Written by Mark Wieczorek (October 2005)
!
!	May 31, 2006 -  Modified so that the Condon-Shortley phase is never used, regardless of 
!			what the default value of CSPHASE_DEFAULT is.
!	June 23, 2007 - Completely rewritten. Can now use 4 different types of grids, and expanded
!			to third order.
!	April 2009 -	Changed the order of the gridtypes
!
!	Copyright (c) 2005-2007, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: MakeGrid2D, MakeGridGLQ, MakeGridDH, NGLQSH, PreCompute

	implicit none
	
	real*8, intent(out) ::	geoid(:,:)
	real*8, intent(in) ::	cilm(:,:,:), r0pot, GM, r, PotRef, omega
	integer, intent(in) :: lmax, order, gridtype
	integer, intent(in), optional :: lmax_calc
	integer, intent(out) :: nlat, nlong
	real*8, intent(in), optional ::	interval, a, f
	real*8 ::	pi, r_ex, lat
	integer :: 	l, nlat1, nlong1, lmax_comp, astat, n, i, astat1, astat2
	real*8, allocatable :: grida(:,:), gridb(:,:), gridc(:,:), gridd(:,:), zero(:), w(:), qq(:,:), pp(:,:), uu(:,:), &
				cilm1(:,:,:), cilm2(:,:,:)

	if ( (present(f) .and. .not. present(a)) .or. (present(a) .and. .not. present(f)) ) then
		print*, "Error --- MakeGeoidGrid"
		print*, "Both F and A must be specified."
		print*, "A, ", present(a)
		print*, "F, ", present(f)
		stop
	endif

	if (gridtype < 1 .or. gridtype > 4) then
		print*, "Error --- MakeGeoidGrid"
		print*, "GRIDTYPE must be either (1) Cartesian 2D, (2) Gauss-Legendre quadrature,"
		print*, "(3) N by N Driscoll and Healy, or (4) N by 2N Driscoll and Healy."
		print*, "Input value = ", gridtype
		stop
	endif
	
	if (gridtype == 4 .and. .not. present(interval)) then
		print*, "Error --- MakeGeoidGrid"
		print*, "If GRIDTYPE = 4 (2D Cartesian), the optional parameter INTERVAL"
		print*, "must be specified."
		stop
	endif
	
	if (order > 3 .or. order < 1) then
		print*, "Error --- MakeGeoidGrid"
		print*, "ORDER must be 1, 2, or 3."
		stop
	endif

	
	if (gridtype == 1) then
		nlong = 2*lmax + 1
		nlat = NGLQSH(lmax)	! nlat = lmax+1
	elseif (gridtype == 2) then
		nlat = 2*lmax+2
		nlong = nlat
	elseif (gridtype == 3) then
		nlat = 2*lmax+2
		nlong = 2*nlat
	elseif (gridtype == 4) then
		nlat = 180.0/interval + 1
		nlong = 360./interval + 1
	endif
		
	if (size(geoid(:,1)) < nlat .or. size(geoid(1,:)) < nlong ) then
		print*, "Error --- MakeGeoidGrid"
		print*, "GEOID must be dimensioned as (180/INTERVAL+1, 360/INTERVAL+1) where INTERVAL is ", interval
		print*, "Input array has dimension ", size(geoid(:,1)), size(geoid(1,:))
		stop
	elseif (size(cilm(:,1,1)) < 2) then
		print*, "Error --- MakeGeoidGrid"
		print*, "CILM must be dimensioned as (2, *, *)."
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	endif

	if (present(lmax_calc)) then
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1, lmax_calc)
	else
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1)
	endif
	
	if (gridtype == 1) then
		allocate(zero(lmax+1), stat = astat)
		if (astat /=0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array ZERO, ", astat
			stop
		endif
		allocate(w(lmax+1), stat = astat)
		if (astat /=0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array W, ", astat
			stop
		endif
		call PreCompute(lmax, zero, w, norm=1, csphase=1)
	endif
	
	allocate(cilm1(2,lmax+1,lmax+1), stat = astat1)
	allocate(cilm2(2,lmax+1,lmax+1), stat = astat2)
	if (astat1 /= 0 .or. astat2 /= 0) then
		print*, "Error --- MakeGeoidGrid"
		print*, "Problem allocating arrays CILM1 and CILM2", astat1, astat2
		stop
	endif

	pi = acos(-1.0d0)

	cilm1(1:2,1:lmax+1, 1:lmax+1) = cilm(1:2,1:lmax+1, 1:lmax+1)
	cilm1(1,1,1) = 1.0d0	! Make sure that the degree-0 term is included

	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	! 
	!	Solve the equation a + b x + c x**2 + d x**3 = 0
	!	for each grid point.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!	
	
	! Create grid A.
	
	do l=0, lmax_comp
		cilm2(1:2,l+1,1:l+1) = (GM/r) * cilm1(1:2,l+1,1:l+1) * (r0pot/r)**l	
	enddo
	! add rotation and subtract reference field
	cilm2(1,1,1) = cilm2(1,1,1) + (omega*r)**2 / 3.0d0 - PotRef
	cilm2(1,3,1) = cilm2(1,3,1) - (omega*r)**2 / (3.0d0*sqrt(5.0d0))
	
	allocate(grida(nlat, nlong), stat = astat)
	if (astat /= 0) then
		print*, "Error --- MakeGeoidGrid"
		print*, "Problem allocating array GRIDA", astat
		stop
	endif
	
	select case(gridtype)
		case(1); call MakeGridGLQ(grida, cilm2, lmax, zero=zero, norm=1, csphase=1, lmax_calc=lmax_comp)
		case(2); call MakeGridDH(grida, n, cilm2, lmax, norm=1, csphase=1, sampling=1, lmax_calc = lmax_comp)
		case(3); call MakeGridDH(grida, n, cilm2, lmax, norm=1, csphase=1, sampling=2, lmax_calc = lmax_comp)
		case(4); call MakeGrid2D(grida, cilm2, lmax_comp, interval, nlat1, nlong1, norm=1, csphase=1)
	end select
	
	! Create Grid B	
	
	do l=0, lmax_comp
		cilm2(1:2,l+1,1:l+1) = -(GM/r**2) * dble(l+1) * cilm1(1:2,l+1,1:l+1) * (r0pot/r)**l
	enddo
	! add rotational terms
	cilm2(1,1,1) = cilm2(1,1,1) + 2.0d0 * r * omega**2 / 3.0d0
	cilm2(1,3,1) = cilm2(1,3,1) - 2.0d0 * r * omega**2 / (3.0d0*sqrt(5.0d0))
	
	allocate(gridb(nlat, nlong), stat = astat)
	if (astat /= 0) then
		print*, "Error --- MakeGeoidGrid"
		print*, "Problem allocating array GRIDB", astat
		stop
	endif
	select case(gridtype)
		case(1); call MakeGridGLQ(gridb, cilm2, lmax, zero=zero, norm=1, csphase=1, lmax_calc=lmax_comp)
		case(2); call MakeGridDH(gridb, n, cilm2, lmax, norm=1, csphase=1, sampling=1, lmax_calc = lmax_comp)
		case(3); call MakeGridDH(gridb, n, cilm2, lmax, norm=1, csphase=1, sampling=2, lmax_calc = lmax_comp)
		case(4); call MakeGrid2D(gridb, cilm2, lmax_comp, interval, nlat1, nlong1, norm=1, csphase=1)
	end select


	! Create Grid C
	
	if (order == 2 .or. order == 3) then
		
		do l=0, lmax_comp
			cilm2(1:2,l+1,1:l+1) = GM/(2.0d0*r**3) * dble(l+1) * dble(l+2) * cilm1(1:2,l+1,1:l+1) * (r0pot/r)**l
		enddo
		! add rotational terms
		cilm2(1,1,1) = cilm2(1,1,1) + omega**2 / 3.0d0
		cilm2(1,3,1) = cilm2(1,3,1) - omega**2 / (3.0d0*sqrt(5.0d0))

		allocate(gridc(nlat, nlong), stat = astat)
		if (astat /= 0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array GRIDC", astat
			stop
		endif
	
		select case(gridtype)
			case(1); call MakeGridGLQ(gridc, cilm2, lmax, zero=zero, norm=1, csphase=1, lmax_calc=lmax_comp)
			case(2); call MakeGridDH(gridc, n, cilm2, lmax, norm=1, csphase=1, sampling=1, lmax_calc = lmax_comp)
			case(3); call MakeGridDH(gridc, n, cilm2, lmax, norm=1, csphase=1, sampling=2, lmax_calc = lmax_comp)
			case(4); call MakeGrid2D(gridc, cilm2, lmax_comp, interval, nlat1, nlong1, norm=1, csphase=1)
		end select
	
	endif
	
	
	! Create Grid D
	
	if (order == 3) then
		
		do l=0, lmax_comp
			cilm2(1:2,l+1,1:l+1) = -GM/(6.0d0*r**4) * dble(l+1) * dble(l+2) * dble(l+3) * &
						cilm1(1:2,l+1,1:l+1) * (r0pot/r)**l
		enddo

		allocate(gridd(nlat, nlong), stat = astat)
		if (astat /= 0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array GRIDD", astat
			stop
		endif
	
		select case(gridtype)
			case(1); call MakeGridGLQ(gridd, cilm2, lmax, zero=zero, norm=1, csphase=1, lmax_calc=lmax_comp)
			case(2); call MakeGridDH(gridd, n, cilm2, lmax, norm=1, csphase=1, sampling=1, lmax_calc = lmax_comp)
			case(3); call MakeGridDH(gridd, n, cilm2, lmax, norm=1, csphase=1, sampling=2, lmax_calc = lmax_comp)
			case(4); call MakeGrid2D(gridd, cilm2, lmax_comp, interval, nlat1, nlong1, norm=1, csphase=1)
		end select
	
	endif
	
	! solve equation
	
	if (order == 1) then
	
		geoid(1:nlat,1:nlong) = -grida(1:nlat,1:nlong) / gridb(1:nlat,1:nlong)
		
	elseif (order == 2) then
		
		geoid(1:nlat, 1:nlong) = (-gridb(1:nlat, 1:nlong) - &
		 		sqrt(gridb(1:nlat, 1:nlong)**2 - 4.0d0*gridc(1:nlat,1:nlong) * &
				grida(1:nlat, 1:nlong)) ) / (2.0d0*gridc(1:nlat,1:nlong))
		
	elseif (order == 3) then
		allocate(qq(nlat, nlong), stat = astat)
		if (astat /= 0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array QQ", astat
			stop
		endif
		allocate(pp(nlat, nlong), stat = astat)
		if (astat /= 0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array PP", astat
			stop
		endif
		allocate(uu(nlat, nlong), stat = astat)
		if (astat /= 0) then
			print*, "Error --- MakeGeoidGrid"
			print*, "Problem allocating array UU", astat
			stop
		endif
		
		pp(1:nlat,1:nlong) = gridb(1:nlat,1:nlong)/gridd(1:nlat,1:nlong) - &
			((gridc(1:nlat,1:nlong)/gridd(1:nlat,1:nlong))**2) / 3.0d0
		qq = grida(1:nlat,1:nlong)/gridd(1:nlat,1:nlong) + &
			2.0d0*((gridc(1:nlat,1:nlong)/gridd(1:nlat,1:nlong))**3) / 27.0d0 - &
			9.0d0*gridc(1:nlat,1:nlong)*gridb(1:nlat,1:nlong)/(gridd(1:nlat,1:nlong)**2) / 27.0d0
		uu(1:nlat,1:nlong) = (qq(1:nlat,1:nlong)/2.0d0 + &
			sqrt((qq(1:nlat,1:nlong)**2)/4.0d0 + (pp(1:nlat,1:nlong)**3)/27.0d0) )**(1.0d0/3.0d0)
		geoid(1:nlat,1:nlong) = pp(1:nlat,1:nlong)/3.0d0/uu(1:nlat,1:nlong) - &
			uu(1:nlat,1:nlong) - gridc(1:nlat,1:nlong)/gridd(1:nlat,1:nlong)/3.0d0
		
		deallocate(qq)
		deallocate(pp)
		deallocate(uu)
		
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Reference geoid to a flattened ellipsoid
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	if (present(a) .and. present(f)) then
		do i = 1, nlat
		
			if (gridtype == 4) then
				lat = 90.0d0 - dble(i-1)*interval
			elseif (gridtype == 1) then
				lat = asin(zero(i)) * 180.0d0 / pi
			elseif (gridtype == 2 .or. gridtype == 3) then
				lat = 90.0d0 - 180.0d0 * dble(i-1)/dble(nlat)
			endif
			
			r_ex = a**2 * (1.0d0 + tan(lat*pi/180.0d0)**2) / &
			(1.0d0  + tan(lat*pi/180.0d0)**2 / (1.0d0 - f)**2 )
			r_ex = sqrt(r_ex)

			geoid(i,1:nlong) = geoid(i,1:nlong) + r - r_ex
		enddo
	endif
		
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Clean up
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if (gridtype == 2) then
		deallocate(zero)
		deallocate(w)
	endif
	deallocate(grida)
	deallocate(gridb)
	if (order == 2) deallocate(gridc)
	if (order == 3) then
		deallocate(gridc)
		deallocate(gridd)
	endif
	
	deallocate(cilm1)
	deallocate(cilm2)
	

end subroutine MakeGeoidGrid
